hub/src/XChainHubSrc.sol

```diff
@@ -0,0 +1,364 @@
+//   ______
+//  /      \
+// /$$$$$$  | __    __  __    __   ______
+// $$ |__$$ |/  |  /  |/  \  /  | /      \
+// $$    $$ |$$ |  $$ |$$  \/$$/ /$$$$$$  |
+// $$$$$$$$ |$$ |  $$ | $$  $$<  $$ |  $$ |
+// $$ |  $$ |$$ \__$$ | /$$$$  \ $$ \__$$ |
+// $$ |  $$ |$$    $$/ /$$/ $$  |$$    $$/
+// $$/   $$/  $$$$$$/  $$/   $$/  $$$$$$/
+//
+// auxo.fi
+
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.12;
+
+/// @dev delete before production commit!
+import "@std/console.sol";
+
+import {IERC20} from "@oz/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@oz/token/ERC20/utils/SafeERC20.sol";
+
+import {Ownable} from "@oz/access/Ownable.sol";
+import {IERC20} from "@oz/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@oz/token/ERC20/utils/SafeERC20.sol";
+import {Pausable} from "@oz/security/Pausable.sol";
+
+import {IVault} from "@interfaces/IVault.sol";
+import {IHubPayload} from "@interfaces/IHubPayload.sol";
+import {IStrategy} from "@interfaces/IStrategy.sol";
+
+import {XChainHubStorage} from "@hub/XChainHubStorage.sol";
+import {XChainHubEvents} from "@hub/XChainHubEvents.sol";
+
+import {LayerZeroApp} from "@hub/LayerZeroApp.sol";
+import {LayerZeroAdapter} from "@hub/LayerZeroAdapter.sol";
+import {IStargateReceiver} from "@interfaces/IStargateReceiver.sol";
+import {IStargateRouter} from "@interfaces/IStargateRouter.sol";
+
+/// @title XChainHub
+/// @notice extends the XChainBase with Stargate and LayerZero contracts for src and destination chains
+/// @dev Expect this contract to change in future.
+abstract contract XChainHubSrc is
+    Pausable,
+    LayerZeroAdapter,
+    XChainHubStorage,
+    XChainHubEvents
+{
+    using SafeERC20 for IERC20;
+
+    constructor(address _stargateRouter) {
+        stargateRouter = IStargateRouter(_stargateRouter);
+    }
+
+    /// ------------------------
+    /// Single Chain Functions
+    /// ------------------------
+
+    /// @notice approve a strategy to withdraw tokens from the hub
+    /// @dev call before withdrawing from the strategy
+    /// @param _strategy the address of the XChainStrategy on this chain
+    /// @param underlying the token
+    /// @param _amount the quantity to approve
+    function approveWithdrawalForStrategy(
+        address _strategy,
+        IERC20 underlying,
+        uint256 _amount
+    ) public onlyOwner {
```

Comment:
it cannot be onlyOwner if called internally

Reply:
It would be better to make it internal, as far as I can see this is only used once which makes me think we don't need a separate function at all.

Reply:
it's used as part of the workflow:

Step 1: Receive tokens into Hub
Step 2: Approve withdrawal from hub
Step 3: Withdraw tokens

I will split into a public ownable function and add an internal non-ownable function